examples/scan_iterator.py

```diff
@@ -0,0 +1,106 @@
+            found = len(bd.name or "") > n or len(ad.local_name or "") > n
+            print(f"Found{' it' if found else ''} {bd!r} with {ad!r}")
+            if found:
+                break
+
+
+async def deterministic_cleanup(run_without_explicit_cleanup: bool = False):
+    # If you want to use async iterator again immediately after it is done,
+    # ensure that the previous one is closed and resources released.
+    # Cleanup of asynchronous generators is a tricky case since they can
+    # be cleaned up by multiple cases at once (unrolling as the generator
+    # finishes, cancellation, or closing as the generator is garbage
+    # collected), the order of which is implementation dependent.
+    # The proper way to address this is not to rely on implicit cleanup,
+    # but await aclose() method or use contextlib.aclosing().
+
+    # Helper method to avoid code duplication, could also be used in `normal_usage()`,
+    # but that would make the `normal_usage()` example less clear.
+    async def iterate_devices(
+        iterator: AsyncIterator[Tuple[BLEDevice, AdvertisementData]], n: int
+    ):
+        print(f"Scanning for {n} devices...")
+        async for bd, ad in iterator:
+            print(f"{n}. {bd!r} with {ad!r}")
+            n -= 1
+            if n == 0:
+                break
+
+    async with BleakScanner() as scanner:
+        print("\nUsing the same scanner async iterator immediately after it is done...")
+
+        if hasattr(contextlib, "aclosing"):  # New in version 3.10
+            if not run_without_explicit_cleanup:
+                print("Using contextlib.aclosing() to ensure cleanup...")
+
+            if run_without_explicit_cleanup:
+                await iterate_devices(scanner.observe(), 2)
+            else:
+                async with contextlib.aclosing(scanner.observe()) as observer:
+                    await iterate_devices(observer, 2)
+
+            if run_without_explicit_cleanup:
+                await iterate_devices(scanner.observe(), 3)
+                await iterate_devices(scanner.observe(), 4)
+            else:
+                async with contextlib.aclosing(scanner.observe()) as observer:
+                    await iterate_devices(observer, 3)
+                    await iterate_devices(observer, 4)
+
+        else:
+            if not run_without_explicit_cleanup:
+                print("Using AsyncIterator.aclose() to ensure cleanup...")
+
+            observer = scanner.observe()
+            await iterate_devices(observer, 2)
+            if not run_without_explicit_cleanup:
+                await observer.aclose()
+
+            observer = scanner.observe()
+            await iterate_devices(observer, 3)
+            if not run_without_explicit_cleanup:
+                await observer.aclose()
+
+        await iterate_devices(scanner.observe(), 6)
+        # This one can be cleaned implicitly, as we do not care about reusing the same scanner anymore
```

Comment:
Looking at this now, `run_without_explicit_cleanup` makes this example too messy. I will replace it with a simple comment and let the users experiment by themselves.